tinydraw_test.go

```diff
@@ -0,0 +1,220 @@
+		return err
+	default:
+		if err != nil {
+			return err
+		}
+	}
+
+	referenceImage, err := png.Decode(referenceFile)
+	if err != nil {
+		return err
+	}
+
+	sx, sy := d.Size()
+
+	differenceDetected := false
+
+	for y := 0; y < int(sy); y++ {
+		for x := 0; x < int(sx); x++ {
+			refPixel := referenceImage.At(x, y)
+			testPixel := d.RGBA.At(x, y)
+
+			refR, refG, refB, refA := refPixel.RGBA()
+			testR, testG, testB, testA := testPixel.RGBA()
+
+			if refR != testR || refG != testG || refB != testB || refA != testA {
+				differenceDetected = true
+				d.RGBA.Set(x, y, color.RGBA{255, 0, 0, 255}) //Mark the pixel as bad for debugging
+			}
+		}
+	}
+
+	if differenceDetected {
+		diffFilePath := filepath.Join("testdata", "DIFF_"+name+".png")
+		saveImgFile, err := os.Create(diffFilePath)
+		if err != nil {
+			return err
+		}
+
+		err = png.Encode(saveImgFile, d.RGBA)
+		return fmt.Errorf("Image contains differences see %v" + diffFilePath)
+	}
+
+	return nil
+}
+
+// Display sends the buffer (if any) to the screen.
+func (d *testDisplay) Display() error {
+	return nil
+}
+
+func TestFilledTriangleClockMinuteHand(t *testing.T) {
+
+	black := color.RGBA{0, 0, 0, 255}
+	unfilledDisplay := newTestDisplay()
+	filledDisplay := newTestDisplay()
+
+	x := int16(160)
+	y := int16(120)
+
+	const (
+		hourRadius   = 55
+		minuteRadius = 95
+		hourWidth    = 6
+		minuteWidth  = 5
+	)
+
+	testTime, _ := time.Parse("15:04:05", "10:02:25")
+	for i := 0; i < 60; i++ {
+		// Draw the clock hands.
+		minuteAngle := float64(testTime.Minute()) / 60 * 2 * math.Pi
+		mx, my := math.Sincos(minuteAngle)
+
+		x0 := x - int16(minuteWidth*my)
+		y0 := y - int16(minuteWidth*mx)
+		x1 := x + int16(minuteWidth*my)
+		y1 := y + int16(minuteWidth*mx)
+		x2 := x + int16(minuteRadius*mx)
+		y2 := y - int16(minuteRadius*my)
+
+		FilledTriangle(
+
+			&filledDisplay,
+			x0, y0,
+			x1, y1,
+			x2, y2,
+			black)
+
+		Triangle(
+			&unfilledDisplay,
+			x0, y0,
+			x1, y1,
+			x2, y2,
+			black)
+
+		testTime = testTime.Add(1 * time.Minute)
+	}
+
+	err := compareWithReference("TestFilledTriangleClockMinuteHand", filledDisplay)
+	if err != nil {
+		t.Errorf("Fail %v", err)
+	}
+
+	err = compareWithReference("TestTriangleClockMinuteHand", unfilledDisplay)
+	if err != nil {
+		t.Errorf("Fail %v", err)
+	}
+}
+
+type triangleTest struct {
+	x0 int16
+	y0 int16
+	x1 int16
+	y1 int16
+	x2 int16
+	y2 int16
+}
+
+func TestTriangles(t *testing.T) {
+
+	tests := []triangleTest{
+		{160, 0, 80, 240, 213, 120},   //Basic triangle with top and bottom half
+		{160, 0, 213, 240, 80, 120},   //Basic triangle with top and bottom half
+		{80, 50, 160, 50, 120, 200},   //No top
+		{80, 200, 160, 200, 120, 50},  //No Bottom
+		{80, 200, 160, 200, 213, 200}, //Horizontal line
+		{80, 80, 80, 120, 80, 200},    //Vertical line
+
+		{160, 115, 160, 125, 255, 120}, //Too Wide
+		{160, 116, 160, 124, 254, 129}, //Fill extends left outside of triangle
+		{165, 120, 155, 120, 160, 215}, //Too Wide
+	}
```

Comment:
If you write the following, it will be easier to check which testdata is which.

```go
tests := []struct{
    expected string
    input triangleText
}{
    {"TestTriangles_0.png", {160, 0, 80, 240, 213, 120}},
    ...
}
```

Reply:
I added the name to the struct and used the name to generate the reference image file name.